src/app/directives/float-only.directive.ts

```diff
@@ -1,23 +1,72 @@
-import { Directive, HostListener } from '@angular/core';
+import { Directive, HostListener, ElementRef } from '@angular/core';
 
 @Directive({
   selector: '[appFloatOnly]'
 })
 export class FloatOnlyDirective {
 
-  constructor() { }
+  constructor(private el: ElementRef) { }
 
   @HostListener('keypress', ['$event']) onKeyPress(event: KeyboardEvent) {
-    const allowedChars = /[0-9.]/;
     const inputChar = String.fromCharCode(event.charCode);
+    const inputValue = this.el.nativeElement.value;
+    const cursorPosition = this.el.nativeElement.selectionStart;
+
+    this.checkPositiveSignPosition(event, inputValue, cursorPosition);
+    this.checkNegativeSignPosition(event, inputValue, cursorPosition);
 
-    if (!allowedChars.test(inputChar)) {
+    // Check if any character is being entered before '-' or '+'
+    if (inputValue[cursorPosition] === '-' || inputValue[cursorPosition] === '+') {
       event.preventDefault();
     }
 
     // Allow only one decimal point
     if (inputChar === '.' && (event.target as HTMLInputElement).value.includes('.')) {
       event.preventDefault();
     }
+
+    if (event) {
+      return this.isVaildFloat(event);
+    }
+  }
+
+  isVaildFloat(event) {
+    const allowedChars = /[0-9.]/;
+    const inputChar = String.fromCharCode(event.charCode);
+    const isAllowedKeys = ['Backspace', 'Tab', 'ArrowLeft', 'ArrowRight', 'Delete', 'e', 'E', '-', '+'].includes(event.key);
+    const isScientificNotation = event.key.toLowerCase() === 'e';
+    const hasAlreadyScientificNotation = this.el.nativeElement.value.includes('e') || this.el.nativeElement.value.includes('E');
+    if (allowedChars.test(inputChar) || isAllowedKeys) {
```

Comment:
Here we have regex check for  /[0-9.]/; Cannot we have same for int?
or
event.key here?

Reply:
Yes, we can have same also.

> Can we have a directive number-only and have common code there; Override that for int and float?

Yes, we can try in that way also but to keep these separate seems better.